Actions logging

danil-topc opened the pull request:
Implement logic to log committed services actions to database

- Add postgres db
- Page with actions
- Retry actions
- Rocket actions logging

matejak commented on src/web/.env.example:
```diff
@@ -1,7 +1,7 @@
 FLASK_APP=autoapp.py
 FLASK_ENV=development
-DATABASE_URL="sqlite:////tmp/dev.db"
+DATABASE_URL="postgresql://postgres:postgres@localhost:5432"
```
Is there some hard reason why to change the backend away from sqlite?

danil-topc replied:
It is possible to treat it as a plain string, but I would have to dump/load it each time manually, and no search possibility. And we would still have to switch to some DBMS, because sqlite is not the best thing for production, I believe.

matejak replied:
I would like the DB platform to be `sqlalchemy`, i.e. DB backend agnostic in the long term. However, I am OK with the current state until we figure out how to structure the data.

danil-topc replied:
hm, my comment wasn't posted or was deleted somehow, will repeat it. 
We're still using sqlalchemy with postgresql. Postgresql brings only one dependency - json field, but it also can be replaced with text field if needed.

matejak replied:
Yes, I agree, but it is not possible any longer to choose a different backend. Although it is common for most projects, TEAP is so lightweight that it is IMO possible to aim for it to be DB-agnostic without sacrificing anything.

matejak replied:
Also please have `DB_HOST`, `DB_USER` and `DB_PASSWORD` as separate config items. That way, the `.env` file will play nicely with docker-compose.

matejak commented on src/web/backend/actions/api.py:
```diff
@@ -0,0 +1,39 @@
+from flask import Blueprint
+from flask_restful import Api, Resource, reqparse
```
We need this updated in Python requirements.

danil-topc replied:
forgot to commit it, added

matejak commented on src/web/local.yml:
```diff
@@ -0,0 +1,18 @@
+      POSTGRES_PASSWORD: postgres
+
+  app:
+    build: .
```
This can work, but let be inspired by this:

```
services:

  db:
    image: "postgres:alpine"
    restart: always
    environment:
      POSTGRES_USER: postgres
      POSTGRES_PASSWORD: postgres
    networks:
      postgres:

  app:
    build: .
    ports:
      - "8080:8080"
    image: teap
    depends_on:
      - db
    networks:
      postgres:

networks:
  postgres:
```

danil-topc replied:
done, also added separate env variables for postgres user, password

matejak commented on src/web/backend/rocket_chat/utils.py:
```diff
@@ -24,8 +29,32 @@ def rocket(self):
+                success = False
+                message = str(e)
+            # TODO: what to do with password in create_user method?
+            filtered_kwargs = {key: value for key, value in kwargs.items() if key != 'password'}
```
This is a good catch - we may want to store only a partial amount of data, so some input may be required upon retry.

danil-topc commented:
Matej, thanks for review! 
Added changes to env file, docker-compose. Working on the rest of changes.

> When attempting to test things out, I got an error stating (psycopg2.ProgrammingError) relation "action" does not exist. The funny thing is that the table action certainly exists in the DB. Any ideas?

If that table exists - it's strange and shouldn't happen. I could assume you didn't run migrations, but then table wouldn't exist.
Tested it on fresh install, works fine. To apply migrations for db and create all relations, run:
`docker-compose -f local.yml run --rm app flask db migrate`

danil-topc commented:
@matejak 
> The Action class connects actual action and details of the DB storage. Let's concentrate on the former - any real-world action has indeed a name, return status, message, kwargs and kwargs that are missing due to filtering.

I'm agree to make some abstract Action class, with abstract methods so every separate action will inherit from this, but why not to leave sqlalchemy db model in this scheme to connect to db and have a data class with it's data manipulation methods (like execute), and to not store all the data as blob ?
It's usual practice to keep business logic with database orm models.

matejak commented:
> ... why not to leave sqlalchemy db model in this scheme to connect to db and have a data class with it's data manipulation methods (like execute), and to not store all the data as blob?

The DB interaction can be moved to a subclass - when you have the ORM model and acual object behavior code in one place, it creates noise that makes it more difficult to learn what's going on.
My motivation for this is to aim for simplicity that would reduce the amount of code as well as the cost of future maintenance by all sorts of coders that may not be that qualified and familiar with ORM practices.

This feature has to exist in some form in the project, but the first iteration of implementation has to cut quite some corners. As I have pointed towards, this feature is about saving actions for display, about being able to retry them after a relatively short period of time, and then about basically forgetting them, as security auditing has to happen on client-side (i.e. LDAP server, Nextcloud server etc.) anyway.

I am sorry that I have not stressed sooner that we expect the project to be approached by those volunteers with various level of expertise - this basically means that if a code construct is somewhat advanced or complicates testing, it is something that we have to take into account and push for complexity reduction. Saving a blob may not be adhering to normal form practices, but it is not super wrong either, and anybody looking into the code will understand that.

matejak commented:
I have tried the code and I observed following things:

1. I have an impression that the database got populated after I ran `db upgrade`, as `db migrate` didn't do anything. Is it possible?
2. When I have added an entint.org user to a division (namely `hello` to `Legal`), I got a super huge traceback.

danil-topc commented:
>I have an impression that the database got populated after I ran db upgrade, as db migrate didn't do anything. Is it possible?

Yep, sorry, didn't notice that. `flask db migrate` creates migration for db, `flask db upgrade` executes it.

> When I have added an entint.org user to a division (namely hello to Legal), I got a super huge traceback.

Cannot reproduce it, works locally. Could you provide me with traceback, please ?

danil-topc commented:
@matejak added class with business logic and all attributes for Action, and inherited orm model from it.

matejak commented:
I have tried, and now, it was DB-related as well:

```
sqlalchemy.exc.ProgrammingError: (psycopg2.ProgrammingError) column "event_name" of relation "action" does not exist
LINE 1: INSERT INTO action (event_name, timestamp, _data, message, s...
                            ^
[SQL: 'INSERT INTO action (event_name, timestamp, _data, message, status) VALUES (%(event_name)s, %(timestamp)s, %(_data)s, %(message)s, %(status)s) RETURNING action.id'] [paramete
rs: {'event_name': 'invite_user_to_channel', 'timestamp': datetime.datetime(2019, 8, 20, 20, 30, 16, 454802), '_data': '{"rocket_channel": "zupT7jJhmscFm6zTa", "rocket_user": "KoHMYf69RT5WJ7
t57"}', 'message': None, 'status': True}] (Background on this error at: http://sqlalche.me/e/f405)
```

Also please resolve conflicts.

danil-topc commented:
> I have tried, and now, it was DB-related as well:

Sorry, didn't add updated migration. I replaced old migration with new one, to not make a lot of migrations inside one PR, so you should start from fresh db.

> Also please resolve conflicts.
Resolved. Also I noticed that `__init__.py` (which was added in #31) in web root directory breaks flask cli and project launch. I removed it, it shouldn't break anything.

matejak commented:
When I click on the actions view, I get a not very helpful traceback ending with

```
PM Exception on /api/actions/ [GET]
...
TypeError: Object of type datetime is not JSON serializable
```

Please test your code using the docker setup, which should make sure that it works on a non-developer environment. When examining the DB, something got saved there though.
Also please get rid of that postgres-specific setup, JSON column datatype and make it a string, so the setup can be tested on one container - it will make everything much simpler.

danil-topc commented:
@matejak removed postgres and compose file, replaced postgres with sqlite and tested it with docker setup. 
However I wasn't able to reproduce the issue with serializing datetime object. There is only one field with datetime - timestamp, but marshmallow handles it's serialization. But may be I'm missing siomething.

matejak commented:
I have had two issues with the PR:

1. The DB URL has to be absolute. While the DB upgrade works with the relative URL, the code itself is not able to find the database and throws schema errors that don't make sense.
2. I guess that the Marshmallow dependency is somehow fragile, as newer versions throw this datetime vs JSON bug. When installed against `prod.txt` deps, it works.

However, the code works, so as soon as you resolve the conflict, it can be merged, congratulations!

danil-topc commented:
Thanks for review! I rebased the code.

Regarding the issues:
> The DB URL has to be absolute. While the DB upgrade works with the relative URL, the code itself is not able to find the database and throws schema errors that don't make sense.

Checked it, you right. Flask-migrations command and flask context looks for a db file in different places: one is in the project root, and another in `backend/`. Will check why it happens.

> I guess that the Marshmallow dependency is somehow fragile, as newer versions throw this datetime vs JSON bug. When installed against prod.txt deps, it works.

Yep, they recently had major upgrade of stable version without backward compatibility. I will add major versions to `prod.in` requirements file.

matejak commented:
Great, thanks!
